research/05-problem-ana.tex

```diff
@@ -1,10 +1,127 @@
+outside of the context of a REPL, for example when the language is a
+DSL. It is clear therefore that a different approach should be
+considered.
+
+\begin{figure}[htbp]
+  \centering
+  \includegraphics[width=0.5\textwidth]{menu-actions.png}
+  \caption{A menu action for the paplj language defined using
+    Spoofax. The bottom window shows the menu definition, the top
+    window shows a program written in paplj.}
+  \label{fig:menu-actions}
+\end{figure}
+
+One possible solution is to allow for language specific configurations
+that are loaded and used during the generation process of a REPL for
+that language. This is similar to something that currently exists in
+Spoofax: when one creates a language, there is the possibility to
+define menu buttons and bind them to Stratego strategies
+(\cref{sec:orgheadline5}). For example, one can define a menu button
+``Run'' to run the expression that is currently selected (see
+\cref{fig:menu-actions}), or a ``Desugar'' button to desugar the
+program that is currently open. In the same way, one may define a
+``list-modules'' command for a REPL which is bound to a Stratego
+strategy.
+
+\subsubsection{Redefining terms bound to names}
+\label{sec:redef-cont-bound}
+When the user is prototyping methods inside the REPL, he or she would
+probably want to be able to redefine that method to be of a different
+implementation. However, this poses a similar problem as in the
+previous section, because for some languages it may not be possible
+nor desirable to do so outside of the context of a REPL. Thus
+requiring the language designer to extend the language with such
+abilities is again an inadequate solution.
+
+One could propose the same solution as in the previous section, namely
+to define an additional command for redefining a class or
+method. However, in this case this is not an adequate solution either,
+since binding terms to names is a common feature among programming
+languages. The fact that Spoofax comes with the NaBL DSL for
+specifying name binding rules illustrates this. Name binding is in
+fact the very feature which enables one to create abstractions.
+
+A possible and more adequate solution for redefining a term bound to a
+name, is to allow the user to give the name and the new term. The name
+can then be used to find the old term, so that it can be replaced with
+the new one.
+
+\subsubsection{Partially replacing the AST of the current program}
+When given an identifier to bind a new term to, the leaves in the AST
+corresponding to the identifier have to be replaced by the new terms.
+This implies that the REPL needs to able parse single terms that might
+not be valid programs, and then update the AST of the program accordingly.
+This requirement has certain effects when it comes to the build process
+used in the REPL.
+
+First of all, the REPL should always hold on to the AST representation of the
+executing program, in order to find the old term in the current program.
```

Comment:
So executing == current. Perhaps you can change this to `such that it is able to find the old term`.